system/mp/mpcomm.cpp

```diff
@@ -2945,46 +2949,102 @@ class CCommunicator: public ICommunicator, public CInterface
         return channel->verifyConnection(tm,true);
     }
 
-    bool verifyAll(bool duplex, unsigned timeout)
+    bool verifyAll(bool duplex, unsigned totalTimeout, unsigned perConnectionTimeout)
     {
-        CriticalBlock block(verifysect);
-        CTimeMon tm(timeout);
+        CTimeMon totalTM(totalTimeout);
+
+        std::vector<std::future<bool>> threads;
+        Semaphore sem;
+        sem.signal(std::thread::hardware_concurrency());
+        bool abort = false;
+        
+        auto verifyConnWithConnect = [&](unsigned rank, unsigned timeout)
         {
-            ForEachNodeInGroup(rank,*group) {
-                bool doverify;
-                if (duplex)
-                    doverify = (rank!=myrank);
-                else if ((rank&1)==(myrank&1))
-                    doverify = (myrank>rank);
-                else
-                    doverify = (myrank<rank);
-                if (doverify) {
+            CTimeMon tm(timeout);
+            CriticalBlock block(verifysect);
+            Owned<CMPChannel> channel = getChannel(rank);
+            return channel->verifyConnection(tm, true);    
```

Comment:
This routine has a CS in it which may serialize execution.

Reply:
yes, in previous draft of this PR, I had a CS per rank,
but in fact originally the CS was held at the start of verifyAll, with some nasty logic to periodically yield if it failed. That was because the CS was global and multiple communicators (e.g. belonging to different channels) could clash and deadlock each other.
The CS is now inside the communicator and it is correct that it is held for the duration of verifyAll, to exclude other calls to verifyAll or disconnect.
(and that means that the threads no longer block each other)

Reply:
I still do not see new threads or any parallelism

Reply:
there's no CS now. So not sure what you mean, can you elaborate?

Reply:
Doesn't the f.get() block ?

Reply:
Should the f.get() be a short wait_for and future_status check is ready ?

Reply:
Shouldn't be necessary, by the time it's in the loop calling f.get() the last N will be running (they will have gained a slot), and will be in the threadedVerifyConnectFunc. They must timeout after approx the overall timeout amount or per connection timeout if shorter.

We need to wait for them to finish either way before leaving verifyAll, so I think f.get() is good.

Reply:
It may be related to std::launch::async ?  See other comment about this.

Reply:
Unrelated, but also if we could eliminate the connectsect CS held for connect() it might offer a potential speedup.